drl/nn/nn.py

```diff
@@ -0,0 +1,526 @@
+		Applies the type of pooling as selected by the user. The default pooling type used is max_pooling.
+
+			:param input_layer: Layer that is fed as input for applying pooling
+			:param name: Optional Parameter. Allows you to name the layer.
+			:return: Pooling applied to the input layer.
+		"""
+		if self.options.POOLING_TYPE == 'max_pooling':
+			return tf.nn.max_pool(input_layer,
+			                      ksize=[1, self.options.POOLING_KSIZE, self.options.POOLING_KSIZE, 1],
+			                      strides=[1, self.options.POOLING_STRIDE, self.options.POOLING_STRIDE, 1],
+			                      padding='SAME', name=name)
+		elif self.options.POOLING_TYPE == 'avg_pooling':
+			return tf.nn.avg_pool(input_layer,
+			                      ksize=[1, self.options.POOLING_KSIZE, self.options.POOLING_KSIZE, 1],
+			                      strides=[1, self.options.POOLING_STRIDE, self.options.POOLING_STRIDE, 1],
+			                      padding='SAME', name=name)
+		else:
+			color_print("ERROR: Please select a pooling operation that is supported by this library!", color='red',
+			            mode='bold')
+			print()
+			color_print(
+				"This library supports the following pooling functions. You can access them by using 'Pooling.YOUR_DESIRED_POOLING_TYPE",
+				color='blue')
+			print_table([Pooling.get_all_pooling()], ['Supported Pooling Types'], indentation='center', color='blue')
+			exit()
+
+	def batch_norm(self, input_layer, decay=0.999, epsilon=1e-3, name=None):
+		"""
+		Batch Normalization wrapper.
+		Original implementation: https://r2rt.com/implementing-batch-normalization-in-tensorflow.html
+
+			:param input_layer: Input layer to which we will apply BN
+			:param decay: Decay for the moving average. Reasonable values for decay are close to 1.0.
+			:param epsilon: Small float added to variance to avoid dividing by zero.
+			:param name: Optional Parameter. Allows you to name the layer.
+			:return: Batch normalized layer output.
+		"""
+		scale = tf.Variable(tf.ones([input_layer.get_shape()[-1]]))
+		beta = tf.Variable(tf.zeros([input_layer.get_shape()[-1]]))
+		pop_mean = tf.Variable(tf.zeros([input_layer.get_shape()[-1]]), trainable=False)
+		pop_var = tf.Variable(tf.ones([input_layer.get_shape()[-1]]), trainable=False)
+
+		def train():
+			batch_mean, batch_var = tf.nn.moments(input_layer, [0])
+			train_mean = tf.assign(pop_mean,
+			                       pop_mean * decay + batch_mean * (1 - decay))
+			train_var = tf.assign(pop_var,
+			                      pop_var * decay + batch_var * (1 - decay))
+			with tf.control_dependencies([train_mean, train_var]):
+				return tf.nn.batch_normalization(input_layer,
+				                                 batch_mean, batch_var, beta, scale, epsilon, name=name)
+
+		def test():
+			return tf.nn.batch_normalization(input_layer,
+			                                 pop_mean, pop_var, beta, scale, epsilon, name=name)
+
+		return tf.cond(self.is_training, lambda: train(), lambda: test())
```

Comment:
Probably the reason why it is not working without lambda is because you pass the called function, e.g. look at this example:

`tf.cond(self.is_training, train(), test())`
What will happen is, the `train` and `test` function are called and the results (which is `tf.nn.batch_normalization(...)`) are passed into the `tf.cond` function, E.g. lets say `result = train()` and now `result` is passed into `tf.cond`, but you cannot perform `result()`. 

Now take the this example:
`tf.cond(self.is_training, train, test)`
In this example the functions `train` and `test` it self is passed into the `tf.cond` function and within the `tf.cond` function some code will execute them by calling `train()` or `test()` at some point. E.g. `train` is passed into `tf.cond` and when needed the function is executed to return `result = train()`.

The reason why `lamda: train()` works is because you construct a new function definition which is passed into `tf.cond`. You can see it as this `new_func = lambda: train()` and `new_func` is passed into `tf_cond`. If you would pass `new_func()` you would get the same problem as the before.

I hope this is clear, i have noticed that passing functions into other functions is a really powerful tool in Python, so it is good to know how it works.